### ISSUE TITLE
5-2 셀프그레이드 관련

### ISSUE THREAD

**Issue by shmsw25** [snu-sf-class/PP2015f_TA#87]
안녕하세요 조교님.
5-2 셀프그레이드에

(define safe-car
  (lambda (l)
    (case-list
     (lambda () 'empty-list)
     (lambda (h t) h)
     l))
  )

이부분에 관련된 질문입니다.
case-list의 첫 두 parameter는 둘의 return값이 같아야 합니다. 그런데 첫 인자의 return값은 symbol이고 두번째 인자의 return 같은 tau로 둘이 달라질 수 있는 것 같습니다. 그냥 'empty-list를 tau라고 하는건가요...?

**Comment by kim-yoonseung** [comment]
말씀하신대로 case-list의 리턴 값의 타입은 같도록 정의되어있습니다.

safe-car는 그저 채점의 편의를 위해 임시로 만든 함수라서 저렇게 되어있습니다.

조금 더 엄밀히 하려면 리턴 타입이 tau+unit 값을 가져야 하는데, 이러면 5-3의 해답이 되어버리므로 임시로 저렇게 하였습니다. 또는 'empty-list 대신 (raise "error: list is empty") 와 같이 exception을 내는 방법도 있겠네요.

저의 생각에는 case-list의 구현을 어떤 방식으로 하셔도 safe-car가 '저의 의도'대로 잘 동작할 것 같았습니다. 혹시 문제의 소지가 있을까요?

**Comment by shmsw25** [comment]
아니요 의도대로 돌아가는 것 같습니다! 그게 5-3 과제에서 저도 저거 비슷한 함수를 만들었는데, 그렇게 하니 주석을 달 때 약간 꼬이게 되어서요... 과제에서는 저런 함수를 만들면 안되나요?!

**Comment by kim-yoonseung** [comment]
사실 테스트에서 잘 동작한다면 괜찮습니다.

primitive 타입을 제외하면 Racket 문법에서 따로 타입을 정해주는 것이 아니므로, 본인이 새로 타입을 정의했다고 주장하셔도 됩니다.

예를 들어 list는 '()과 pair라는 두 가지 방법을 통해 만들어지는데, pair와 '()는 딱히 관련이 없죠.

그리고 각자가 이 문제에서 sum과 list 타입의 원소를 임의로 정의한 것과 같이, 본인이 AAA라는 타입을 `{0,'AAA}` 라고 정의하였다고 주장하실 수 있겠습니다.
